File: backend/services/trading_algorithms.py
import pandas as pd
import numpy as np
from typing import Dict, List, Tuple, Optional
from datetime import datetime, timedelta

class TradingAlgorithms:
    """Trading algorithms for forex analysis"""
    
    def __init__(self):
        self.name = "TradingAlgorithms"
    
    @staticmethod
    def calculate_sma(prices: List[float], period: int) -> List[float]:
        """Calculate Simple Moving Average"""
        if len(prices) < period:
            return [None] * len(prices)
        
        sma = []
        for i in range(len(prices)):
            if i < period - 1:
                sma.append(None)
            else:
                avg = sum(prices[i-period+1:i+1]) / period
                sma.append(avg)
        return sma
    
    @staticmethod
    def calculate_rsi(prices: List[float], period: int = 14) -> List[float]:
        """Calculate Relative Strength Index"""
        if len(prices) < period + 1:
            return [None] * len(prices)
        
        deltas = []
        for i in range(1, len(prices)):
            deltas.append(prices[i] - prices[i-1])
        
        gains = [max(delta, 0) for delta in deltas]
        losses = [-min(delta, 0) for delta in deltas]
        
        rsi = [None]  # First value is always None
        
        # Calculate initial RSI
        avg_gain = sum(gains[:period]) / period
        avg_loss = sum(losses[:period]) / period
        
        if avg_loss == 0:
            rsi.append(100)
        else:
            rs = avg_gain / avg_loss
            rsi.append(100 - (100 / (1 + rs)))
        
        # Calculate subsequent RSI values
        for i in range(period + 1, len(prices)):
            avg_gain = (avg_gain * (period - 1) + gains[i-1]) / period
            avg_loss = (avg_loss * (period - 1) + losses[i-1]) / period
            
            if avg_loss == 0:
                rsi.append(100)
            else:
                rs = avg_gain / avg_loss
                rsi.append(100 - (100 / (1 + rs)))
        
        return rsi
    
    @staticmethod
    def calculate_bollinger_bands(prices: List[float], period: int = 20, std_dev: int = 2) -> Tuple[List[float], List[float], List[float]]:
        """Calculate Bollinger Bands (upper, middle, lower)"""
        if len(prices) < period:
            none_list = [None] * len(prices)
            return none_list, none_list, none_list
        
        middle_band = TradingAlgorithms.calculate_sma(prices, period)
        upper_band = []
        lower_band = []
        
        for i in range(len(prices)):
            if i < period - 1 or middle_band[i] is None:
                upper_band.append(None)
                lower_band.append(None)
            else:
                period_prices = prices[i-period+1:i+1]
                std = np.std(period_prices)
                upper_band.append(middle_band[i] + (std_dev * std))
                lower_band.append(middle_band[i] - (std_dev * std))
        
        return upper_band, middle_band, lower_band
    
    def sma_crossover_signal(self, prices: List[float], short_period: int = 10, long_period: int = 30) -> Optional[str]:
        """Generate SMA crossover trading signals"""
        if len(prices) < long_period:
            return None
        
        short_sma = self.calculate_sma(prices, short_period)
        long_sma = self.calculate_sma(prices, long_period)
        
        # Check last two values for crossover
        if len(short_sma) < 2 or short_sma[-1] is None or short_sma[-2] is None:
            return None
        if long_sma[-1] is None or long_sma[-2] is None:
            return None
        
        # Bullish crossover (short SMA crosses above long SMA)
        if short_sma[-2] <= long_sma[-2] and short_sma[-1] > long_sma[-1]:
            return 'BUY'
        
        # Bearish crossover (short SMA crosses below long SMA)
        if short_sma[-2] >= long_sma[-2] and short_sma[-1] < long_sma[-1]:
            return 'SELL'
        
        return None
    
    def rsi_signal(self, prices: List[float], period: int = 14, oversold: int = 30, overbought: int = 70) -> Optional[str]:
        """Generate RSI trading signals"""
        rsi_values = self.calculate_rsi(prices, period)
        
        if len(rsi_values) < 2 or rsi_values[-1] is None:
            return None
        
        current_rsi = rsi_values[-1]
        
        if current_rsi <= oversold:
            return 'BUY'  # Oversold, potential buy signal
        elif current_rsi >= overbought:
            return 'SELL'  # Overbought, potential sell signal
        
        return None
    
    def bollinger_bands_signal(self, prices: List[float], period: int = 20, std_dev: int = 2) -> Optional[str]:
        """Generate Bollinger Bands trading signals"""
        upper, middle, lower = self.calculate_bollinger_bands(prices, period, std_dev)
        
        if len(prices) < 2 or upper[-1] is None or lower[-1] is None:
            return None
        
        current_price = prices[-1]
        
        # Price touches lower band - potential buy signal
        if current_price <= lower[-1]:
            return 'BUY'
        
        # Price touches upper band - potential sell signal
        if current_price >= upper[-1]:
            return 'SELL'
        
        return None
    
    def generate_combined_signal(self, prices: List[float]) -> Dict:
        """Generate combined signal from all algorithms"""
        signals = {
            'sma': self.sma_crossover_signal(prices),
            'rsi': self.rsi_signal(prices),
            'bollinger': self.bollinger_bands_signal(prices),
            'consensus': None
        }
        
        # Simple consensus logic
        buy_signals = sum(1 for signal in signals.values() if signal == 'BUY')
        sell_signals = sum(1 for signal in signals.values() if signal == 'SELL')
        
        if buy_signals >= 2:
            signals['consensus'] = 'BUY'
        elif sell_signals >= 2:
            signals['consensus'] = 'SELL'
        
        return signals

File: server/trading/TradingAlgorithms.js
class TradingAlgorithms {
  constructor() {
    this.priceHistory = new Map(); // symbol -> array of prices
    this.maxHistory = 200; // Keep last 200 data points
  }

  addPriceData(symbol, price, timestamp) {
    if (!this.priceHistory.has(symbol)) {
      this.priceHistory.set(symbol, []);
    }
    
    const history = this.priceHistory.get(symbol);
    const mid = (price.bid + price.ask) / 2;
    
    history.push({
      timestamp: timestamp || new Date(),
      price: mid,
      high: mid + (price.ask - price.bid) / 2,
      low: mid - (price.ask - price.bid) / 2,
      volume: price.volume || 0
    });
    
    // Keep only recent history
    if (history.length > this.maxHistory) {
      history.splice(0, history.length - this.maxHistory);
    }
  }

  getPriceHistory(symbol, periods = null) {
    const history = this.priceHistory.get(symbol) || [];
    if (periods) {
      return history.slice(-periods);
    }
    return history;
  }

  // Simple Moving Average
  calculateSMA(symbol, period) {
    const history = this.getPriceHistory(symbol, period);
    if (history.length < period) {
      return null;
    }
    
    const sum = history.reduce((acc, item) => acc + item.price, 0);
    return sum / period;
  }

  // SMA Crossover Strategy
  getSMACrossoverSignal(symbol, shortPeriod = 10, longPeriod = 50) {
    const shortSMA = this.calculateSMA(symbol, shortPeriod);
    const longSMA = this.calculateSMA(symbol, longPeriod);
    
    if (!shortSMA || !longSMA) {
      return { signal: 'hold', confidence: 0, reason: 'insufficient_data' };
    }

    // Get previous SMAs to detect crossover
    const history = this.getPriceHistory(symbol);
    if (history.length < longPeriod + 1) {
      return { signal: 'hold', confidence: 0, reason: 'insufficient_history' };
    }

    // Calculate previous SMAs
    const prevHistory = history.slice(0, -1);
    const prevShortSum = prevHistory.slice(-shortPeriod).reduce((acc, item) => acc + item.price, 0);
    const prevLongSum = prevHistory.slice(-longPeriod).reduce((acc, item) => acc + item.price, 0);
    const prevShortSMA = prevShortSum / shortPeriod;
    const prevLongSMA = prevLongSum / longPeriod;

    // Detect crossover
    const currentCross = shortSMA > longSMA;
    const prevCross = prevShortSMA > prevLongSMA;
    
    let signal = 'hold';
    let confidence = 0;
    let reason = 'no_crossover';

    if (currentCross && !prevCross) {
      // Golden cross - bullish signal
      signal = 'buy';
      confidence = Math.min(((shortSMA - longSMA) / longSMA) * 100, 1.0);
      reason = 'golden_cross';
    } else if (!currentCross && prevCross) {
      // Death cross - bearish signal
      signal = 'sell';
      confidence = Math.min(((longSMA - shortSMA) / shortSMA) * 100, 1.0);
      reason = 'death_cross';
    }

    return {
      signal,
      confidence,
      reason,
      shortSMA,
      longSMA,
      crossover: currentCross !== prevCross
    };
  }

  // Relative Strength Index (RSI)
  calculateRSI(symbol, period = 14) {
    const history = this.getPriceHistory(symbol, period + 1);
    if (history.length < period + 1) {
      return null;
    }

    const gains = [];
    const losses = [];

    for (let i = 1; i < history.length; i++) {
      const change = history[i].price - history[i - 1].price;
      if (change > 0) {
        gains.push(change);
        losses.push(0);
      } else {
        gains.push(0);
        losses.push(Math.abs(change));
      }
    }

    const avgGain = gains.reduce((sum, gain) => sum + gain, 0) / period;
    const avgLoss = losses.reduce((sum, loss) => sum + loss, 0) / period;

    if (avgLoss === 0) {
      return 100; // All gains, RSI = 100
    }

    const rs = avgGain / avgLoss;
    const rsi = 100 - (100 / (1 + rs));

    return rsi;
  }

  // RSI Strategy
  getRSISignal(symbol, period = 14, overbought = 70, oversold = 30) {
    const rsi = this.calculateRSI(symbol, period);
    
    if (rsi === null) {
      return { signal: 'hold', confidence: 0, reason: 'insufficient_data', rsi: null };
    }

    let signal = 'hold';
    let confidence = 0;
    let reason = 'neutral';

    if (rsi > overbought) {
      signal = 'sell';
      confidence = Math.min((rsi - overbought) / (100 - overbought), 1.0);
      reason = 'overbought';
    } else if (rsi < oversold) {
      signal = 'buy';
      confidence = Math.min((oversold - rsi) / oversold, 1.0);
      reason = 'oversold';
    }

    return {
      signal,
      confidence,
      reason,
      rsi,
      overbought,
      oversold
    };
  }

  // Bollinger Bands
  calculateBollingerBands(symbol, period = 20, standardDeviations = 2) {
    const history = this.getPriceHistory(symbol, period);
    if (history.length < period) {
      return null;
    }

    const prices = history.map(item => item.price);
    const sma = prices.reduce((sum, price) => sum + price, 0) / period;
    
    const squaredDifferences = prices.map(price => Math.pow(price - sma, 2));
    const variance = squaredDifferences.reduce((sum, diff) => sum + diff, 0) / period;
    const stdDev = Math.sqrt(variance);

    return {
      middle: sma,
      upper: sma + (standardDeviations * stdDev),
      lower: sma - (standardDeviations * stdDev),
      bandwidth: (2 * standardDeviations * stdDev) / sma * 100
    };
  }

  // Bollinger Bands Strategy
  getBollingerBandsSignal(symbol, period = 20, standardDeviations = 2) {
    const bands = this.calculateBollingerBands(symbol, period, standardDeviations);
    const history = this.getPriceHistory(symbol, 1);
    
    if (!bands || history.length === 0) {
      return { signal: 'hold', confidence: 0, reason: 'insufficient_data', bands: null };
    }

    const currentPrice = history[history.length - 1].price;
    let signal = 'hold';
    let confidence = 0;
    let reason = 'within_bands';

    if (currentPrice > bands.upper) {
      signal = 'sell';
      confidence = Math.min((currentPrice - bands.upper) / (bands.upper - bands.middle), 1.0);
      reason = 'above_upper_band';
    } else if (currentPrice < bands.lower) {
      signal = 'buy';
      confidence = Math.min((bands.lower - currentPrice) / (bands.middle - bands.lower), 1.0);
      reason = 'below_lower_band';
    }

    return {
      signal,
      confidence,
      reason,
      bands,
      currentPrice,
      position: currentPrice > bands.upper ? 'above_upper' : 
               currentPrice < bands.lower ? 'below_lower' : 'within_bands'
    };
  }

  // Combined Signal Analysis
  getCombinedSignal(symbol, options = {}) {
    const {
      smaShort = 10,
      smaLong = 50,
      rsiPeriod = 14,
      rsiOverbought = 70,
      rsiOversold = 30,
      bbPeriod = 20,
      bbStdDev = 2
    } = options;

    const smaSignal = this.getSMACrossoverSignal(symbol, smaShort, smaLong);
    const rsiSignal = this.getRSISignal(symbol, rsiPeriod, rsiOverbought, rsiOversold);
    const bbSignal = this.getBollingerBandsSignal(symbol, bbPeriod, bbStdDev);

    const signals = [smaSignal, rsiSignal, bbSignal];
    const validSignals = signals.filter(s => s.confidence > 0);

    if (validSignals.length === 0) {
      return {
        signal: 'hold',
        confidence: 0,
        reason: 'no_valid_signals',
        components: { sma: smaSignal, rsi: rsiSignal, bb: bbSignal }
      };
    }

    // Calculate weighted average
    const buySignals = validSignals.filter(s => s.signal === 'buy');
    const sellSignals = validSignals.filter(s => s.signal === 'sell');

    let finalSignal = 'hold';
    let confidence = 0;

    if (buySignals.length > sellSignals.length) {
      finalSignal = 'buy';
      confidence = buySignals.reduce((sum, s) => sum + s.confidence, 0) / buySignals.length;
    } else if (sellSignals.length > buySignals.length) {
      finalSignal = 'sell';
      confidence = sellSignals.reduce((sum, s) => sum + s.confidence, 0) / sellSignals.length;
    }

    return {
      signal: finalSignal,
      confidence: Math.min(confidence, 1.0),
      reason: 'combined_analysis',
      components: { sma: smaSignal, rsi: rsiSignal, bb: bbSignal },
      buyCount: buySignals.length,
      sellCount: sellSignals.length,
      totalSignals: validSignals.length
    };
  }
}

module.exports = TradingAlgorithms;